src/coprocessor/statistics/analyze.rs

```diff
@@ -73,27 +77,48 @@ impl<S: Snapshot> AnalyzeContext<S> {
         Ok(res_data)
     }
 
-    // handle_index is used to handle `AnalyzeIndexReq`,
+    // batch_handle_index is used to handle `AnalyzeIndexReq`,
     // it would build a histogram and count-min sketch of index values.
-    fn handle_index(
+    fn batch_handle_index(
         req: AnalyzeIndexReq,
-        scanner: &mut IndexScanExecutor<TiKVStorage<SnapshotStore<S>>>,
+        scanner: &mut BatchIndexScanExecutor<TiKVStorage<SnapshotStore<S>>>,
     ) -> Result<Vec<u8>> {
         let mut hist = Histogram::new(req.get_bucket_size() as usize);
         let mut cms = CmSketch::new(
             req.get_cmsketch_depth() as usize,
             req.get_cmsketch_width() as usize,
         );
-        while let Some(row) = scanner.next()? {
-            let row = row.take_origin()?;
-            let (bytes, end_offsets) = row.data.get_column_values_and_end_offsets();
-            hist.append(bytes);
-            if let Some(c) = cms.as_mut() {
-                for end_offset in end_offsets {
-                    c.insert(&bytes[..end_offset])
+
+        let mut is_drained = false;
+        let mut batch_size = BATCH_INITIAL_SIZE;
+        while !is_drained {
+            use std::ops::Index;
+
+            let batch = scanner.next_batch(batch_size);
+            is_drained = batch.is_drained?;
+
+            for logical_row in batch.logical_rows {
+                let mut bytes = vec![];
+                for col in batch.physical_columns.as_slice() {
+                    let buffer_vec = col.raw();
+                    let data = buffer_vec.index(logical_row);
+                    bytes.extend_from_slice(data);
```

Comment:
Looks like the behaviour is scanning data from row, convert to column, and here convert it back to row?

Reply:
> Looks like the behaviour is scanning data from row, convert to column, and here convert it back to row?

I think a more suitable description is scanner scanned a matrix, then here get every row from matrix.